### ISSUE TITLE
Coscheduling: When the resource is sufficient, the pod is always Pending

### ISSUE THREAD

**Issue by liiitleboy** [kubernetes-sigs/scheduler-plugins#114]
apiVersion: "batch/v1"
kind: "Job"
metadata:
  namespace: test-ns
  name: tf-work-1
spec:
  template:
    metadata:
      labels:
        pod-group.scheduling.sigs.k8s.io/name: tf-work
        pod-group.scheduling.sigs.k8s.io/min-available: "2"
    spec:
      restartPolicy: "OnFailure"
      containers:
        - image: myimage
          name: tf-work-1
          command: [ "/bin/sh","-c","top -b" ]
          resources:
            limits:
              cpu: "6"
              memory: 8Gi
            requests:
              cpu: "6"
              memory: 8Gi

apiVersion: "batch/v1"
kind: "Job"
metadata:
  namespace: test-ns
  name: tf-work-2
spec:
  template:
    metadata:
      labels:
        pod-group.scheduling.sigs.k8s.io/name: tf-work
        pod-group.scheduling.sigs.k8s.io/min-available: "2"
    spec:
      restartPolicy: "OnFailure"
      containers:
        - image: myimage
          name: tf-work-2
          command: [ "/bin/sh","-c","top -b" ]
          resources:
            limits:
              cpu: "6"
              memory: 8Gi
            requests:
              cpu: "6"
              memory: 8Gi

`E1124 12:15:02.731405       1 factory.go:601] Error scheduling test-ns/tf-work-1-qbtz6: rejected by "coscheduling" at prefilter: less than minAvailable; retrying
E1124 12:15:02.735282       1 scheduler.go:627] error selecting node for pod: rejected by "coscheduling" at prefilter: less than minAvailable
E1124 12:15:02.735520       1 factory.go:601] Error scheduling test-ns/tf-work-1-qbtz6: rejected by "coscheduling" at prefilter: less than minAvailable; retrying
E1124 12:15:02.735564       1 scheduler.go:627] error selecting node for pod: rejected by "coscheduling" at prefilter: less than minAvailable
E1124 12:15:24.119796       1 factory.go:601] Error scheduling test-ns/tf-work-2-9t8jg: pod "tf-work-2-9t8jg" rejected while waiting at permit: rejected due to timeout after waiting 10s at plugin coscheduling; retrying
E1124 12:15:34.124308       1 factory.go:601] Error scheduling test-ns/tf-work-2-9t8jg: pod "tf-work-2-9t8jg" rejected while waiting at permit: rejected due to timeout after waiting 10s at plugin coscheduling; retrying
E1124 12:16:39.419128       1 factory.go:601] Error scheduling test-ns/tf-work-1-qbtz6: pod "tf-work-1-qbtz6" rejected while waiting at permit: rejected due to timeout after waiting 10s at plugin coscheduling; retrying
`

**Comment by denkensk** [comment]
preFilter is enhanced feature to reduce the overall scheduling time for the whole group. It will check the total number of pods belonging to the same PodGroup. If the total number is less than minAvailable, the pod will reject in preFilter, then the scheduling cycle will interrupt. And the preFilter is user selectable according to the actual situation of users. If the minAvailable of PodGroup is relatively small, for example less than 5, you can disable this plugin. But if the minAvailable of PodGroup is relatively large, please enable this plugin to reduce the overall scheduling time.

So You can try to disable the prefilter to solve this problem.

**Comment by denkensk** [comment]
And It is recommended to add your cluster information

**Comment by liiitleboy** [comment]
> preFilter is enhanced feature to reduce the overall scheduling time for the whole group. It will check the total number of pods belonging to the same PodGroup. If the total number is less than minAvailable, the pod will reject in preFilter, then the scheduling cycle will interrupt. And the preFilter is user selectable according to the actual situation of users. If the minAvailable of PodGroup is relatively small, for example less than 5, you can disable this plugin. But if the minAvailable of PodGroup is relatively large, please enable this plugin to reduce the overall scheduling time.
> 
> So You can try to disable the prefilter to solve this problem.

thanks,how can I disable the prefilter?
K8S 1.18
请问这玩意怎么禁用，集群信息是指什么

**Comment by denkensk** [comment]
Replace config like this
```
apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
leaderElection:
  leaderElect: false
clientConnection:
  kubeconfig: "REPLACE_ME_WITH_KUBE_CONFIG_PATH"
profiles:
- schedulerName: default-scheduler
  plugins:
    queueSort:
      enabled:
        - name: Coscheduling
      disabled:
        - name: "*"
    permit:
      enabled:
        - name: Coscheduling
    reserve:
      enabled:
        - name: Coscheduling
    postBind:
      enabled:
        - name: Coscheduling
```

**Comment by denkensk** [comment]
Update for 1.18

```yaml
apiVersion: kubescheduler.config.k8s.io/v1alpha2
kind: KubeSchedulerConfiguration
leaderElection:
  leaderElect: false
clientConnection:
  kubeconfig: "REPLACE_ME_WITH_KUBE_CONFIG_PATH"
profiles:
- schedulerName: default-scheduler
  plugins:
    queueSort:
      enabled:
        - name: Coscheduling
      disabled:
        - name: "*"
    preFilter:
      enabled:
        - name: Coscheduling
    permit:
      enabled:
        - name: Coscheduling
    unreserve:
      enabled:
        - name: Coscheduling
```

**Comment by liiitleboy** [comment]
> Update for 1.18
> 
> ```yaml
> apiVersion: kubescheduler.config.k8s.io/v1alpha2
> kind: KubeSchedulerConfiguration
> leaderElection:
>   leaderElect: false
> clientConnection:
>   kubeconfig: "REPLACE_ME_WITH_KUBE_CONFIG_PATH"
> profiles:
> - schedulerName: default-scheduler
>   plugins:
>     queueSort:
>       enabled:
>         - name: Coscheduling
>       disabled:
>         - name: "*"
>     preFilter:
>       enabled:
>         - name: Coscheduling
>     permit:
>       enabled:
>         - name: Coscheduling
>     unreserve:
>       enabled:
>         - name: Coscheduling
> ```


I use command ：
kubectl replace -f x.yaml 
>error: unable to recognize "x.yaml": no matches for kind "KubeSchedulerConfiguration" in version 

kubectl get KubeSchedulerConfiguration
>error: the server doesn't have a resource type "KubeSchedulerConfiguration"

can you give me any help?

**Comment by TomLan42** [comment]
I have experienced a similar problem when I am testing the coschedulnig plugin with Kubemark. 

I prepared 2 hollow-nodes, each with 8 GPU. Then I submit several dozens of jobs sequentially (i.e. 60 jobs in total, submit every 15 seconds, each job runs for around 30 seconds), each requiring 1-8 GPU cards. I would work fine when there were sufficient GPU resources and every pod in a job could get allocated without pending. 

When there were not enough GPU resources for the incoming jobs, those jobs will start pending, which is as expected. However, when the running jobs succeeded and released enough resources for the pending jobs, no pending jobs would start. All the jobs would still pending even if there were enough resources for at least one of them because 1-8 always < 16.

I suspect this is a typical deadlocking situation that happens in the 'Permit' stage. And I indeed observed the interleaving of pods of different podgroup waiting in the logs. 

It is reported in the design doc that the QueueSort Plugins ensures the Pod belonging to the same PodGroup will be placed back to back to avoid deadlock (I am not sure if this is also the method of the latest PodGroup CRD based coscheduling). However, the QueueSort Plugins only have an effect on the ActiveQ. 

Imagine the situations where some pods of a podgroup are in the Permit stage, while others may be thrown into BackoffQ or UnschedulableQ because of failed scheduling due to insufficient resources. When the later pods will be flushed back to ActiveQ is kind of 'random'. When the pods in the Permit stage expire and requeue (or thrown into BackoffQ/UnschedulableQ), they may never have the chance to 'reunite' with their peers in the ActiveQ back-to-back. The result is, there is always only a part of the pods of each podgroup are competing in the reserve stage thus deadlock happens. 

This is just suspicion based on what I observe in my experiment and what I read about the coscheudling plugins and kube-scheduler SchedulilngQueue Mechanism. I worry that the BackoffQ and UnschedulableQ mechanism will somehow invalidate some of the assumptions the coscheduling plugins implies.
